src/ophyd_async/epics/pvi/pvi.py

```diff
@@ -0,0 +1,323 @@
+                )
+
+            if isinstance(sub_entries, dict) and (
+                len(sub_entries) != 1 or pre_defined_device_vector
+            ):
+                sub_device: Union[DeviceVector, Device] = DeviceVector()
+
+                for sub_split_number, sub_entry in sub_entries.items():
+                    if sub_entry.is_pvi_table:  # If the entry isn't a signal
+                        if (
+                            sub_common_device_type
+                            and get_origin(sub_common_device_type) == DeviceVector
+                        ):
+                            sub_common_device_type = get_args(sub_common_device_type)[0]
+                        sub_entry.device = (
+                            sub_common_device_type()
+                            if sub_common_device_type
+                            else Device()
+                        )
+                        self.initialize_device(
+                            sub_entry, common_device_type=sub_common_device_type
+                        )
+                    else:  # entry is a signal
+                        signal_type = (
+                            get_args(sub_common_device_type)[0]
+                            if sub_common_device_type
+                            else None
+                        )
+                        sub_entry.device = _pvi_mapping[sub_entry.access](
+                            signal_type, *sub_entry.values
+                        )
+                    assert isinstance(sub_device, DeviceVector)
+                    sub_device[sub_split_number] = sub_entry.device
+            else:
+                if isinstance(sub_entries, dict):
+                    sub_device = (
+                        sub_common_device_type() if sub_common_device_type else Device()
+                    )
+                    assert list(sub_entries) == [1]
+                    sub_entries[1].device = sub_device
+                    self.initialize_device(
+                        sub_entries[1], common_device_type=sub_common_device_type
+                    )
+                else:  # entry is a signal
+                    signal_type = (
+                        get_args(sub_common_device_type)[0]
+                        if sub_common_device_type
+                        else None
+                    )
+                    sub_device = _pvi_mapping[sub_entries.access](
+                        signal_type, *sub_entries.values
+                    )
+
+            setattr(entry.device, sub_name, sub_device)
+
+        # Check that all predefined devices are present
+        if common_device_type:
+            _verify_common_blocks(entry, common_device_type)
+
+
+def _sim_common_blocks(device: Device, stripped_type: Optional[Type] = None):
+    device_t = stripped_type or type(device)
+    for sub_name, sub_device_t in get_type_hints(device_t).items():
+        if sub_name in ("_name", "parent"):
+            continue
```

Comment:
Can we do `issubclass(sub_device_t, Device)` instead of this?

Reply:
This won't work because `sub_device_t` isn't strictly a class, it could be
`typing.Optional[ophyd_async.core.device.Device]` or `DeviceVector[SignalR[int]]`. We parse out the `Device` type later in the loop.